scripts/generate-docs/custom_doc.go

```diff
@@ -0,0 +1,281 @@
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"sort"
+	"strings"
+	"text/template"
+
+	"gopkg.in/yaml.v2"
+
+	"github.com/pkg/errors"
+)
+
+type customDocEvent struct {
+	filePath    string
+	dataStream  string
+	fileSubPath string
+	doc         eventDoc
+}
+
+type eventOverview struct {
+	Name        string `yaml:"name"`
+	Description string `yaml:"description"`
+}
+
+type eventIdentification struct {
+	Os         []string          `yaml:"os"`
+	DataStream string            `yaml:"data_stream"`
+	Filter     map[string]string `yaml:"filter"`
+}
+
+type eventFieldDetails struct {
+	Description string `yaml:"description"`
+}
+
+type eventFields struct {
+	Endpoint []string                     `yaml:"endpoint"`
+	Details  map[string]eventFieldDetails `yaml:"details"`
+}
+
+type eventDoc struct {
+	Overview       eventOverview       `yaml:"overview"`
+	Identification eventIdentification `yaml:"identification"`
+	Fields         eventFields         `yaml:"fields"`
+}
+
+func findCustomDocFiles(root string) ([]customDocEvent, error) {
+	// Return all the custom doc files under root. Intelligently understand the directory structure
+	// and interpret the data_stream name from it for each file.
+	var customFiles []customDocEvent
+
+	dataStreamRoot := filepath.Clean(filepath.Join(root, "data_stream"))
+	dataStreamNames, err := os.ReadDir(dataStreamRoot)
+	if err != nil {
+		return nil, errors.Wrapf(err, "Failed to list directory %s", dataStreamRoot)
+	}
+	for _, dataStreamName := range dataStreamNames {
+		dataStreamNameRoot := filepath.Join(dataStreamRoot, dataStreamName.Name())
+		err := filepath.WalkDir(dataStreamNameRoot, func(path string, info os.DirEntry, err error) error {
+
+			if !info.IsDir() {
+				cleanPath := filepath.Clean(path)
+				event := customDocEvent{filePath: path, dataStream: dataStreamName.Name(), fileSubPath: cleanPath[len(dataStreamRoot):]}
+				customFiles = append(customFiles, event)
+			}
+			return nil
+		})
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to load data_stream directory %s", dataStreamName.Name())
+		}
+	}
+	return customFiles, err
+}
+
+func loadCustomDocFile(path string) (eventDoc, error) {
+	var result eventDoc
+
+	body, err := ioutil.ReadFile(path)
+	if err != nil {
+		return result, errors.Wrapf(err, "reading file failed (path: %s)", path)
+	}
+
+	err = yaml.UnmarshalStrict(body, &result)
+	if err != nil {
+		return result, errors.Wrapf(err, "parsing yaml failed (path: %s)", path)
+	}
+
+	// do some light validation
+	if result.Overview.Name == "" {
+		return result, fmt.Errorf("missing overview.name in %s", path)
+	}
+	if result.Overview.Description == "" {
+		return result, fmt.Errorf("missing overview.description in %s", path)
+	}
+
+	if len(result.Identification.Os) == 0 {
+		return result, fmt.Errorf("missing identification.os in %s", path)
+	}
+	for _, _os := range result.Identification.Os {
+		low := strings.ToLower(_os)
+		if low != "linux" && low != "macos" && low != "windows" {
```

Comment:
Why not enforce strict casing for consistency?
(And does go have a cleaner "item in list" syntax?)
```suggestion
		if _os != "linux" && _os != "macos" && _os != "windows" {
```

Reply:
Ah. I see that you do enforce consistency later - on output.